### PR TITLE
Create a more compact diff format

### PR DESCRIPTION
**Pull request by oli-obk**
An example showcasing all the possible render combinations

Capture

### CODE REVIEW

**Comment by RalfJung** [comment]
Some "N lines skipped" have "..." after them and some not -- what's up with that?

Is it possible to always print a few lines around where things changed, for context (the way `diff -u` does)?

**Comment by oli-obk** [comment]
> Is it possible to always print a few lines around where things changed, for context (the way `diff -u` does)?

done

@rustbot review

**Comment by RalfJung** [comment]
I think I'm going to have to just trust you on this one, since that `row` function would be a lot of effort to review properly. I don't know the API it interacts with and it has very few comments.

Can you show how an example diff look like now?

File: ui_test/src/diff.rs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                        for char in &diff {
+                            match char {
+                                Left(_) => {}
+                                Right(r) => eprint!("{}", r.to_string().green()),
+                                Both(l, _) => eprint!("{}", l),
+                            }
+                        }
+                        eprintln!();
+                    } else {
+                        eprint!("{}", "~".yellow());
```
Comment (RalfJung) [inline ui_test/src/diff.rs:104]: Please add a comment explaining this. I have never seen a tilde in a diff...

Comment (RalfJung) [reply ui_test/src/diff.rs:104]: Oh is this the "inline diff" case? The comments make it sound like `if let Some(l) = self.prev_left.take()` is the inline diff case, but then the "display both lines" is *within* that branch so the comment seems misleading.

**Comment by oli-obk** [comment]
> I think I'm going to have to just trust you on this one

My plan was to just see how it goes in practice and adjust it if we see problems. Worst case is just some badly rendered diffs

**Comment by oli-obk** [comment]
two random missing lines in a stderr cause:

image

at the two line skip point:

image

when only one line would be skipped:

image

when just some chars change in a line:

image

when you both add and remove chars from a line:

image

when you do larger diffs:

image

the last one is not great... I could make it like "more than half of the chars must be the same"

**Comment by RalfJung** [comment]
Nice!

I think the one thing I am missing is whether `-` or `+` is the reference output vs actual output. Regular diff starts with a header like
```
--- build.rs    2022-07-25 12:43:20.977864910 -0400
+++ Cargo.toml  2022-07-25 12:43:20.977864910 -0400
```
that makes this quite clear; should we have something similar?

**Comment by RalfJung** [comment]
Also, does this always print the first and last line? Not that I object, just a bit surprising. In particular it seems to be the first two lines and the last one line, which is strangely asymmetric?

**Comment by oli-obk** [comment]
continued in https://github.com/oli-obk/ui_test/pull/8

**Pull request closed without merging**